test/hotspot/jtreg/compiler/lib/template_framework/Template.java

```diff
@@ -0,0 +1,599 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.lib.template_framework;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import java.util.List;
+
+/**
+ * The Template Framework allows the generation of code with Templates. The goal is that these Templates are
+ * easy to write, and allow regression tests to cover a larger scope, and to make template based fuzzing easy
+ * to extend.
+ *
+ * <p>
+ * <strong>Motivation:</strong> We want to make it easy to generate variants of tests. Often, we would like to
+ * have a set of tests, corresponding to a set of types, a set of operators, a set of constants, etc. Writing all
+ * the tests by hand is cumbersome or even impossible. When generating such tests with scripts, it would be
+ * preferable if the code generation happens automatically, and the generator script was checked into the code
+ * base. Code generation can go beyond simple regression tests, and one might want to generate random code from
+ * a list of possible templates, to fuzz individual Java features and compiler optimizations.
+ *
+ * <p>
+ * The Template Framework provides a facility to generate code with Templates. Templates are essencially a list
+ * of tokens that are concatenated (i.e. rendered) to a String. The Templates can have "holes", which are
+ * filled (replaced) by different values at each Template instantiation. For example, these "holes" can
+ * be filled with different types, operators or constants. Templates can also be nested, allowing a modular
+ * use of the Templates.
+ *
+ * <p>
+ * <strong>Example:</strong>
+ * The following are snippets from the example test {@code TestAdvanced.java}.
+ * First, we define a template that generates a {@code @Test} method for a given type, operator and
+ * constant generator. We define two constants {@code con1} and {@code con2}, and then use a multiline
+ * string with hashtag {@code #} "holes" that are then replaced by the template arguments and the
+ * {@link #let} definitions.
+ *
```

Comment:
Should we also insert `<p>` between the snippets? Same further down at method comments.

Reply:
Ah nice catch! I put them in at the beginning, but then expanded the documentation and forgot it there.